Update about logo icon (colour) to denote the chain type of the QT instance in About/ Help Message Window/ Dialog

pablomartin4btc opened the pull request:
It fixes https://github.com/bitcoin-core/gui/issues/761.

About logo icon is updated now with the correct colour indicating the chain type of the QT instance that user is running (as the splash screen), plus the title of the about window also is updated with the chain type string as in the main window.
<details>
<summary>Before PR changes screenshot</summary>

![image](https://github.com/bitcoin-core/gui/assets/110166421/63945955-2e29-4dc9-8a2e-2a5567983e9b)
</details>

<details>
<summary>After PR changes screenshot</summary>

![image](https://github.com/bitcoin-core/gui/assets/110166421/b6ce187f-ca81-4c4b-b2f6-a2e3b15e6010)
</details>

<details>
<summary>Current splash screen screenshot</summary>
(this hasn't been touched, just to show consistency with this PR changes)


![image](https://github.com/bitcoin-core/gui/assets/110166421/2a1c329b-baa6-4603-bf20-dac2746682f4)
</details>

Tested it also with:
- mainnet (no command line args after `bitcoin-qt` or `bitcoin-qt -chain=main`), 
- `-regtest` and 
- `-testnet`.

pablomartin4btc committed:
gui: Update about logo icon to denote chain type

Adding the networkStyle parameter to the HelpMessageDialog
creator on utilitydialog, updating all calls where its instance
is being created from bitcoingui.cpp. In the object itself,
use this new parameter object to build the about window title and
set the icon of the about logo widget.

maflcko reviewed (commented):
It may help to show screenshots before and after

pablomartin4btc commented:
> It may help to show screenshots before and after

Thanks for your review @MarcoFalke, the before screenshot is on the issue #761 but I can bring it here if that makes it clearer.

pablomartin4btc commented on src/qt/utilitydialog.cpp:
```diff
@@ -135,6 +137,19 @@ void HelpMessageDialog::on_okButton_accepted()
+void HelpMessageDialog::setAboutWindowTitle(const NetworkStyle *networkStyle)
+{
+    QString aboutTitle = tr("About %1").arg(PACKAGE_NAME);
+    if ((networkStyle) && (Params().GetChainType() != ChainType::MAIN)) aboutTitle.append(" " + networkStyle->getTitleAddText());
```
I didn't want to touch `src/qt/bitcoin.cpp`:

https://github.com/bitcoin-core/gui/blob/db19a7e89d19d97094bcf8753b37619a16b7f033/src/qt/bitcoin.cpp#L568

And I also needed to move where the `networkStyle` is initialised/ set before the `HelpMessageDialog` constructor:

https://github.com/bitcoin-core/gui/blob/db19a7e89d19d97094bcf8753b37619a16b7f033/src/qt/bitcoin.cpp#L607

But I can do it if that's better.

hernanmarino reviewed (approved):
tested ACK

pablomartin4btc commented:
> It may help to show screenshots before and after

@maflcko I've updated the description with your suggestion (copied @furszy's style from his PR #764, I find it less invasive for the reviewer)

MarnixCroes reviewed (approved):
tack 30c235b2e6fd9b7e3634d795cedc171d4d8d0e27

kristapsk commented on src/qt/utilitydialog.cpp:
```diff
@@ -135,6 +137,19 @@ void HelpMessageDialog::on_okButton_accepted()
+
+void HelpMessageDialog::setChainTypeIconOnAboutLogo(const NetworkStyle *networkStyle)
+{
+    const QSize requiredSize(1024,1024);
```
nit
```suggestion
    const QSize requiredSize(1024, 1024);
```

pablomartin4btc replied:
Thanks for reviewing! I'll change it if I have to re-touch it.

pablomartin4btc replied:
done!

BrandonOdiwuor reviewed (commented):
Concept ACK

BrandonOdiwuor reviewed (approved):
Tested ACK 30c235b2e6fd9b7e3634d795cedc171d4d8d0e27

![Screenshot from 2023-12-28 16-34-43](https://github.com/bitcoin-core/gui/assets/15610188/9e70728e-6ee9-4981-9c5a-9797d1a8a170)

hebasto commented:
@GBKS What do you think from a designer's point of view?

GBKS commented:
Definitely more consistent this way.

hebasto commented:
> Not sure. All info displayed on the About page is fixed details about the software itself, not about the current (test)chain.

In that case, shouldn't the logo on the About page be black-toned?

pablomartin4btc commented:
> > Not sure. All info displayed on the About page is fixed details about the software itself, not about the current (test)chain.
> 
> In that case, shouldn't the logo on the About page be black-toned?

I'm less sure about black-toned icons, in that specific case, it's more common usage I guess, users are already used to the orange one.

I think the most notable use case is when you have different instances open and you want to make sure you are standing on the correct one but more importantly the consistency regarding other components such as the splash screen and task-bar icons as pointed by @GBKS.

hebasto commented:
https://github.com/bitcoin-core/gui/pull/762#issuecomment-1946005139:
> Definitely more consistent this way.

Thank you @GBKS for your designer's opinion!

hebasto commented on src/qt/utilitydialog.h:
```diff
@@ -34,6 +36,8 @@ class HelpMessageDialog : public QDialog
 
 private Q_SLOTS:
     void on_okButton_accepted();
+    void setAboutWindowTitle(const NetworkStyle *networkStyle = nullptr);
+    void setChainTypeIconOnAboutLogo(const NetworkStyle *networkStyle = nullptr);
```
Why these two member functions are declared as slots?

pablomartin4btc replied:
Sorry, that's wrong, I meant to put them just on the private section. I'll fix it soon. Thanks!

pablomartin4btc replied:
done!

hebasto reviewed (commented):
If you will touch this branch, mind applying [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/clang-format-diff.py)?

pablomartin4btc commented:
Updates:
<ul><li><details>
<summary>Addressed @hebasto's <a href="https://github.com/bitcoin-core/gui/pull/762#pullrequestreview-2058694015">feedback</a>.</summary>

```
git diff -U0 HEAD~1.. | ./contrib/devtools/clang-format-diff.py -p1 -i -v
Formatting src/qt/bitcoingui.cpp
Formatting src/qt/utilitydialog.cpp
Formatting src/qt/utilitydialog.h

```
</details></li></ul>

- As part of the above, [functions](https://github.com/bitcoin-core/gui/pull/762#discussion_r1602056777) were re-placed in the `private` section (no in slots as mistaken).
- As a consequence of applying the [clang-formatting](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md#clang-format-diffpy), @kristapsk's [feedback](https://github.com/bitcoin-core/gui/pull/762#discussion_r1437190084) got addressed as well.

hebasto commented on src/qt/utilitydialog.cpp:
```diff
@@ -29,18 +31,17 @@
-HelpMessageDialog::HelpMessageDialog(QWidget *parent, bool about) :
-    QDialog(parent, GUIUtil::dialog_flags),
-    ui(new Ui::HelpMessageDialog)
+HelpMessageDialog::HelpMessageDialog(QWidget* parent, bool about, const NetworkStyle* networkStyle) : QDialog(parent, GUIUtil::dialog_flags),
+                                                                                                      ui(new Ui::HelpMessageDialog)
```
Break the long line for better readability:
```suggestion
HelpMessageDialog::HelpMessageDialog(QWidget* parent, bool about, const NetworkStyle* networkStyle)
    : QDialog(parent, GUIUtil::dialog_flags),
      ui(new Ui::HelpMessageDialog)
```
?

pablomartin4btc replied:
Ok, will do.

hebasto commented on src/qt/utilitydialog.cpp:
```diff
@@ -135,6 +136,19 @@ void HelpMessageDialog::on_okButton_accepted()
+void HelpMessageDialog::setAboutWindowTitle(const NetworkStyle* networkStyle)
+{
+    QString aboutTitle = tr("About %1").arg(PACKAGE_NAME);
+    if ((networkStyle) && (Params().GetChainType() != ChainType::MAIN)) aboutTitle.append(" " + networkStyle->getTitleAddText());
```
Putting a non-trivial expression into a separate line can facilitate debugging in the future:
```suggestion
    if (networkStyle && Params().GetChainType() != ChainType::MAIN) {
        aboutTitle.append(" " + networkStyle->getTitleAddText());
    }
```

pablomartin4btc replied:
Ok, I'll update it, thanks!

hebasto reviewed (commented):
Tested 577d500cb1ac4122fba3202c289454dcbe31a194.

On Windows, there remains some inconsistency. Running `bitcoin-qt.exe -signet -about` ends with a dialog with an orange-coloured icon.

pablomartin4btc commented:
> On Windows, there remains some inconsistency. Running `bitcoin-qt.exe -signet -about` ends with a dialog with an orange-coloured icon.

Let me take a look...

pablomartin4btc commented:
> > On Windows, there remains some inconsistency. Running `bitcoin-qt.exe -signet -about` ends with a dialog with an orange-coloured icon.
> 
> Let me take a look...

![image](https://github.com/user-attachments/assets/6ce38de9-a5a1-4d09-adfb-3442139db0aa)

Good catch... trying to get it sorted...

pablomartin4btc commented:
Updates:
- Addressed @hebasto's feedback.
- Fixed the [error message window inconsistency](https://github.com/bitcoin-core/gui/pull/762#issuecomment-2276259183) detected by @hebasto [above](https://github.com/bitcoin-core/gui/pull/762#pullrequestreview-2219281894).

<details>
<summary>Now running <code>qt</code> from the command line with <code>bitcoin-qt.exe -signet -about</code> (or similar command with an invalid argument) will show the correct chain type icon at the top left  of the title bar (this is only happening in Windows at the moment).</summary>

![image](https://github.com/user-attachments/assets/d5dd5959-12c1-4278-9b5d-349cb55a7755)
 
</details>

pablomartin4btc commented:
Rebased.

pablomartin4btc commented on src/qt/bitcoingui.cpp:
```diff
@@ -105,7 +105,7 @@ BitcoinGUI::BitcoinGUI(interfaces::Node& node, const PlatformStyle *_platformSty
 
     rpcConsole = new RPCConsole(node, _platformStyle, nullptr);
-    helpMessageDialog = new HelpMessageDialog(this, false);
+    helpMessageDialog = new HelpMessageDialog(this, false, m_network_style);
```
Done.

pablomartin4btc committed:
gui: Update QMessageBox::Icon with chaintype image

Introducing a new helper GUIUtil::ShowMessageBox() in order to
incorporate the chaintype image in the message box window icon.

pablomartin4btc commented:
_<ins>Updates:</ins>_
- Addressed @laanwj's feedback.
- Reworked the [error message window inconsistency](https://github.com/bitcoin-core/gui/pull/762#issuecomment-2276259183) fix detected by @hebasto [above](https://github.com/bitcoin-core/gui/pull/762#pullrequestreview-2219281894), applying it to most `QMessageBox` possible instances in `bitcoin.cpp` and separating it in the 2nd commit for clarity and in case it needs to be dropped because it's getting out of scope or extended to other files (perhaps a follow-up in a separate PR).
  <details>
  <summary>Message-box window icon behavior is inconsistent across platforms: some X11 window managers and many Wayland compositors ignore message-box icons entirely. With the new <code>GUIUtil::ShowMessageBox()</code> helper, we can instead explicitly set <code>QIcon()</code> (no icon) and avoid constructing a <code>NetworkStyle</code> just for dialogs or checking whether a <code>ChainType</code> was passed.
  <br> <i>(Windows behavior still needs to be re-checked under Qt 6 — to confirm whether the no-icon option is effectively applied.)</i></summary>
    
    ```diff --git a/src/qt/guiutil.cpp b/src/qt/guiutil.cpp
    --- a/src/qt/guiutil.cpp
    +++ b/src/qt/guiutil.cpp
    @@ -1012,7 +1012,6 @@ void ShowModalDialogAsynchronously(QDialog* dialog)
     
     void ShowMessageBox(const QString& message,
                                 QMessageBox::Icon box_icon,
    -                            const NetworkStyle* network_style,
                                 const QString& title)
     {
         QString qTitle = CLIENT_NAME;
    @@ -1021,14 +1020,7 @@ void ShowMessageBox(const QString& message,
     
         mBox.setTextFormat(Qt::PlainText);
   
    -    if (network_style) {
    -        mBox.setWindowIcon(network_style->getTrayAndWindowIcon());
    -    } else if (!gArgs.GetChainTypeString().empty()) {
    -        std::unique_ptr<const NetworkStyle> fallback(NetworkStyle::instantiate(gArgs.GetChainType()));
    -        if (fallback) {
    -            mBox.setWindowIcon(fallback->getTrayAndWindowIcon());
    -        }
    -    }
    +    mBox.setWindowIcon(QIcon());
   
         mBox.exec();
     }
  
    ```
    
    </details>
